lib/vtls/cipher_suite.c

```diff
@@ -0,0 +1,699 @@
+{
+  uint8_t indexes[8] = {0};
+  const char *entry, *cur;
+  const char *nxt = cs_str;
+  const char *end = cs_str + cs_len;
+  char separator = '-';
+  int idx, i = 0;
+  size_t len;
+
+  /* split the cipher string by '-' or '_' */
+  if(strncasecompare(cs_str, "TLS", 3))
+    separator = '_';
+
+  do {
+    if(i == 8)
+      return -1;
+
+    /* determine the length of the part */
+    cur = nxt;
+    for(; nxt < end && *nxt != '\0' && *nxt != separator; nxt++);
+    len = nxt - cur;
+
+    /* lookup index for the part (skip empty string at 0) */
+    for(idx = 1, entry = cs_txt + 1; idx < CS_TXT_LEN; idx++) {
+      size_t elen = strlen(entry);
+      if(elen == len && strncasecompare(entry, cur, len))
+        break;
+      entry += elen + 1;
+    }
+    if(idx == CS_TXT_LEN)
+      return -1;
+
+    indexes[i++] = (uint8_t) idx;
+  } while(nxt < end && *(nxt++) != '\0');
+
+  /* zip the 8 indexes into 48 bits */
+  zip[0] = (uint8_t) (indexes[0] << 2 | (indexes[1] & 0x3F) >> 4);
+  zip[1] = (uint8_t) (indexes[1] << 4 | (indexes[2] & 0x3F) >> 2);
+  zip[2] = (uint8_t) (indexes[2] << 6 | (indexes[3] & 0x3F));
+  zip[3] = (uint8_t) (indexes[4] << 2 | (indexes[5] & 0x3F) >> 4);
+  zip[4] = (uint8_t) (indexes[5] << 4 | (indexes[6] & 0x3F) >> 2);
+  zip[5] = (uint8_t) (indexes[6] << 6 | (indexes[7] & 0x3F));
+
+  return 0;
+}
+
+static int cs_zip_to_str(const uint8_t zip[6],
+                         char *buf, size_t buf_size)
+{
+  uint8_t indexes[8] = {0};
+  const char *entry;
+  char separator = '-';
+  int idx, i, r;
+  size_t len = 0;
+
+  /* unzip the 8 indexes */
+  indexes[0] = zip[0] >> 2;
+  indexes[1] = ((zip[0] << 4) & 0x3F) | zip[1] >> 4;
+  indexes[2] = ((zip[1] << 2) & 0x3F) | zip[2] >> 6;
+  indexes[3] = ((zip[2] << 0) & 0x3F);
+  indexes[4] = zip[3] >> 2;
+  indexes[5] = ((zip[3] << 4) & 0x3F) | zip[4] >> 4;
+  indexes[6] = ((zip[4] << 2) & 0x3F) | zip[5] >> 6;
+  indexes[7] = ((zip[5] << 0) & 0x3F);
+
+  if(indexes[0] == CS_TXT_IDX_TLS)
+    separator = '_';
+
+  for(i = 0; i < 8 && indexes[i] != 0 && len < buf_size; i++) {
+    if(indexes[i] >= CS_TXT_LEN)
+      return -1;
+
+    /* lookup the part string for the index (skip empty string at 0) */
+    for(idx = 1, entry = cs_txt + 1; idx < indexes[i]; idx++) {
+      size_t elen = strlen(entry);
+      entry += elen + 1;
+    }
+
+    /* append the part string to the buffer */
+    if(i > 0)
+      r = msnprintf(&buf[len], buf_size - len, "%c%s", separator, entry);
+    else
+      r = msnprintf(&buf[len], buf_size - len, "%s", entry);
+
+    if(r < 0)
+      return -1;
+    len += r;
+  }
+
+  return 0;
+}
+
+uint16_t Curl_cipher_suite_lookup_id(const char *cs_str, size_t cs_len)
+{
+  size_t i;
+  uint8_t zip[6];
+
+  if(cs_len > 0 && cs_str_to_zip(cs_str, cs_len, zip) == 0) {
+    for(i = 0; i < CS_LIST_LEN; i++) {
+      if(memcmp(cs_list[i].zip, zip, sizeof(zip)) == 0)
+        return cs_list[i].id;
+    }
+  }
+
+  return 0;
+}
+
+static bool is_separator(char c)
```

Comment:
```suggestion
static bool cs_is_separator(char c)
```
and same for callers. To avoid the chance of overlapping with another function with the same name.

Reply:
Sure can do, but is it really necessary though? As this is static function. BTW the exact same function can be found in vtls/sectransp.c and vtls/bearssl.c.

Reply:
It's common failure mode for "Unity" builds where all sources are merged together into a single C file and then compiled.

This update would fix SecureTransport + mbedTLS MultiSSL builds with CMake with Unity mode enabled.

The BearSSL / SecureTransport cases also need to be fixed. Thanks for pointing to this.

Reply:
Oh I did not know about unity builds, seems they were introduced after my last contribution. (I must say I have the same concerns as jay though about it: https://github.com/curl/curl/pull/11095#issuecomment-1541418372 )

Will fix, what is preferred: add a commit or overwrite commit + force push?

I have PR's for BearSSL and SecureTransport lined up for consideration based on this PR if this gets merged.

Reply:
Thanks, no strong perference, adding a commit is fine IMO.

For the BearSSL / SecureTransport I opened #13450 earlier.

Reply:
> no strong perference

Ok, in this case I prefer overwrite + force. Changed code as suggested.